pkg/snapshot/snapshot_file.go

```diff
@@ -171,401 +332,1055 @@ type MilestoneDiffProducerFunc func() (*MilestoneDiff, error)
-	FileHeader
-	// The time at which the snapshot was taken.
-	Timestamp uint32
-	// The count of solid entry points.
-	SEPCount uint64
-	// The count of outputs. This count is zero if a delta snapshot has been read.
+	// Active Protocol Parameter of the ledger milestone index.
+	ProtocolParamsMilestoneOpt *iotago.ProtocolParamsMilestoneOpt
+	// The amount of UTXOs contained within this snapshot.
 	OutputCount uint64
-	// The count of milestone diffs.
-	MilestoneDiffCount uint64
+	// The amount of milestone diffs contained within this snapshot.
+	MilestoneDiffCount uint32
+	// The amount of SEPs contained within this snapshot.
+	SEPCount uint16
 }
 
-// getSnapshotFilesLedgerIndex returns the final ledger index if the given snapshot files would be applied.
-func getSnapshotFilesLedgerIndex(fullHeader *ReadFileHeader, deltaHeader *ReadFileHeader) iotago.MilestoneIndex {
-
-	if fullHeader == nil {
-		return 0
-	}
+func (h *FullSnapshotHeader) ProtocolParameters() (*iotago.ProtocolParameters, error) {
 
-	if deltaHeader == nil {
-		return fullHeader.SEPMilestoneIndex
+	protoParams := &iotago.ProtocolParameters{}
+	if _, err := protoParams.Deserialize(h.ProtocolParamsMilestoneOpt.Params, serializer.DeSeriModeNoValidation, nil); err != nil {
+		return nil, fmt.Errorf("failed to deserialize protocol parameters: %w", err)
 	}
 
-	return deltaHeader.SEPMilestoneIndex
+	return protoParams, nil
 }
 
-// StreamSnapshotDataTo streams a snapshot data into the given io.WriteSeeker.
-// FileHeader.Type is used to determine whether to write a full or delta snapshot.
-// If the type of the snapshot is Full, then OutputProducerFunc must be provided.
-func StreamSnapshotDataTo(writeSeeker io.WriteSeeker, timestamp uint32, header *FileHeader,
-	sepProd SEPProducerFunc, outputProd OutputProducerFunc, msDiffProd MilestoneDiffProducerFunc) (*SnapshotMetrics, error) {
+func writeFullSnapshotHeader(writeSeeker io.WriteSeeker, header *FullSnapshotHeader) (int64, error) {
 
-	if header.Type == Full {
-		switch {
-		case outputProd == nil:
-			return nil, ErrOutputProducerNotProvided
-		case header.TreasuryOutput == nil:
-			return nil, ErrTreasuryOutputNotProvided
-		}
+	if header.Type != Full {
+		return 0, ErrWrongSnapshotType
+	}
+	if header.ProtocolParamsMilestoneOpt == nil {
+		return 0, iotago.ErrMissingProtocolParas
+	}
+	if header.TreasuryOutput == nil {
+		return 0, ErrTreasuryOutputNotProvided
 	}
 
-	var sepsCount, outputCount, msDiffCount uint64
-
-	timeStart := time.Now()
+	var countersFileOffset int64
 
-	// write LS file version and type
-	if _, err := writeSeeker.Write([]byte{header.Version, byte(header.Type)}); err != nil {
-		return nil, fmt.Errorf("unable to write LS version and type: %w", err)
+	// Version
+	// Denotes the version of this file format.
+	if _, err := writeSeeker.Write([]byte{header.Version}); err != nil {
+		return 0, fmt.Errorf("unable to write LS version: %w", err)
 	}
+	increaseOffsets(serializer.OneByte, &countersFileOffset)
```

Comment:
Instead of this, add up the ints returned by `writeSeeker.Write`.
Manually calculating offsets is more error-prone